### ISSUE TITLE
Generalise MonoidK

### ISSUE THREAD

**Issue by kcsongor** [typelevel/cats#1318]
I was looking at various type classes, and this sentence caught my eye: "MonoidK is a universal monoid which operates on kinds." - great, this is a really useful abstraction!
However, I couldn't help but think that this could be generalised further to general monoidal categories, not just the cartesian closed category of scala types.
In general, a monoidal category is a category equipped with an associative bifunctor `F x F -> F` (+ an identity object). Now, MonoidK as it exists today, is specialised to cartesian monoidal categories, i.e. ones in which the the product looks like `(F, F) -> F`, so we can smash two F's into one F. (This specialisation takes place in the type of `combine`, which takes two `F[A]`s).

My proposal is to abstract away from the cartesian product, and write it in terms of arbitrary products, so combine would be something like `Product[F, F] => F`. Then the current formulation would be something like `Cartesian[F, F] => F`, with `type Cartesian[A, B] = (A, B)`. (for identity, we need a `Void` type, so `(Void, A) ~ (A, Void) ~ A`)
Another monoidal category (that is not a cartesian category) is the category of endofunctors. There, `Product` can be thought of as composition of endofunctors, so compose would `Composition[F, F] => F`, which translates to `F[F[A]] => F[A]`.

Why is this a good idea?
I guess it demonstrates that a monad is really _just_ a monoid in the category of endofunctors. (note the type of `compose` when the product is `Composition` - this is just `join`!) Appropriate t-shirts can then be printed without remorse.

PS:
Isn't Monoid just MonoidK[Id]?

**Comment by non** [comment]
@kcsongor Well, the difference is that `Monoid[A]` is bound to `A` whereas `MonoidK[Id]` is still able to operate on any values you have in hand. But you're right, if you had a `MonoidK[F]` you can get a `Monoid[F[A]]` for any particular `A`.

One issue is that `MonoidK[Id]` would have to be something like *choose-left* or *choose-right* which usually aren't the monoids we are most interested in. This ends up being a general problem when comparing universal monoids to non-universal (particular?) monoids: sometimes you want to combine `F[A]` values based on something more than the `F` structure.

As far as the proposal goes -- it's an interesting idea, but I'm uncertain (`MonoidK` is already pretty far afield). The endofunctor example is interesting -- is there a situation you can think of where we'd want to generalize across monoidal categories?

(It would also be possible to leave `MonoidK` where it is but create a whole `MonoidalK` hierarchy somewhere else that explores this connection.)

**Comment by johnynek** [comment]
`Monoid` is not `MonoidK[Id]`. The latter has a combine for all `A`. The former only for some `A`.

**Comment by kcsongor** [comment]
@non @johnynek thanks, I realised slightly after that I was only thinking in terms of the resulting type signature, not how the parameters are quantified, which was an obvious mistake.

**Comment by kcsongor** [comment]
@non I only had a really brief look at the gitter channel earlier, and it struck me that there is a little bit of a misunderstanding - I remember seeing statements like "not all monads are monoids", and that "monads in CT and monads in CS are entirely different concepts".
I understand where these are coming from - it is true that it's hard to see how these relate to each other, but making such statements in my opinion is harmful, because it can convince others that it is the case. And indeed, no matter how hard I look at `Monoid` next to `Monad` in scala, I will not be able to see how on earth could a `Monad` be a `Monoid`, and I might be inclined to make such statements as above.

Another problem here is that it might be unreasonable to tell people to do further reading in category theory in order to better understand the concept, but it might be perfectly fine to just point them to a specific (slightly more obscure, but still scala) source file that exhibits these connections.

Long story short, while having this abstraction might have too much practical use per se, I think it has really good educational value.
On the other hand, if the Monad type class had references to endofunctor categories, with biendofunctors and what-not, it might be even more difficult to digest, so this is not a clear win.
